### DISCUSSION TITLE
Circular import when using `get_asset_key_for_model`

### DISCUSSION THREAD

**Discussion by malow106** [dagster-io/dagster#17812] [Q&A] [answered]
Hi folks, 

I try the new `get_asset_key_for_model` but get an error.
`NameError: name 'my_dbt_assets' is not defined`
My setup :
In _\app\Dagster_proj\assets\__init__.py_, I declare my `my_dbt_assets`
```
@dbt_assets(
    manifest=dbt_manifest_path, dagster_dbt_translator=CustomDagsterDbtTranslator()
)
def my_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource):
    yield from dbt.cli(["build"], context=context).stream()
```
In _\app\Dagster_proj\assets\export\__init__.py_, I try to create a dependency to dbt model repair_log with get_asset_key_for_model
```
from dagster import asset,AssetIn,Output,SourceAsset,MetadataValue,RetryPolicy,FreshnessPolicy
from dagster_dbt import get_asset_key_for_model

@asset(
    ins={
        "repair_log": AssetIn(
            input_manager_key="dwh_io_manager",
            key=get_asset_key_for_model([dbt_assets], "repair_log"),
        )
    },
    compute_kind="Excel",
    freshness_policy=FreshnessPolicy(maximum_lag_minutes=60*6,cron_schedule=config['cron_schedule'] or "0 6 * * *",cron_schedule_timezone="Europe/Paris")
)
def export_repair_log(repair_log):
 ...
```
I have tried to import with :
`from ...assets import my_dbt_assets`
But got this error :
```dagster._core.errors.DagsterImportError: Encountered ImportError: `cannot import name 'my_dbt_assets' from partially initialized module 'Dagster_proj.assets' (most likely due to a circular import) (C:\Users\XXXXXX\Documents\DIS_ETL_Pipeline\app\Dagster_proj\assets\__init__.py)` while importing module Dagster_proj. Local modules were resolved using the working directory `C:\Users\XXXXXX\Documents\DIS_ETL_Pipeline\app`. If another working directory should be used, please explicitly specify the appropriate path using the `-d` or `--working-directory` for CLI based targets or the `working_directory` configuration option for workspace targets.```

Thanks !

**Comment by rexledesma** [comment, marked as answer]
The error message is helpful here:

> Encountered ImportError: `cannot import name 'my_dbt_assets' from partially initialized module 'Dagster_proj.assets' (most likely due to a circular import) 

You have a circular dependency. You can fix this by ensuring your modules aren't importing each other simultaneously. Read up on this here: https://medium.com/brexeng/avoiding-circular-imports-in-python-7c35ec8145ed.

**Reply by malow106** [reply]
Yes, but without this import there is another error 
```
dagster._core.errors.DagsterImportError: Encountered ImportError: `cannot import name 'my_dbt_assets'
```
So how to use `my_dbt_assets` in an other module ?

**Reply by rexledesma** [reply]
Fixing the circular dependency does not mean removing the import. It means that you should restructure your modules so that you can import `my_dbt_assets` in a non-circular way.

I am assuming you have your `Definitions` object in your `assets/init.py` file, alongside your `my_dbt_assets` definition. And in this file, you are importing `assets/external/__init__.py`. If so, this is the cause of your circular import.

If you want to fix this, then you should move your `my_dbt_assets` definition to another file (say `assets/dbt/__init__.py` and let `assets/external/__init__.py` import from that. Then, `assets/init.py` can import from both `assets/dbt/__init__.py` and `assets/external/__init__.py`.
